apis/python/src/tiledbsoma/io/registration/ambient_label_mappings.py

```diff
@@ -0,0 +1,402 @@
+        cls,
+        experiment_uri: Optional[str] = None,
+        *,
+        obs_field_name: str = "obs_id",
+        var_field_name: str = "var_id",
+        # XXX CTX/CFG
+    ) -> Self:
+        """Factory method to compute label-to-SOMA-join-ID mappings for a single SOMA experiment in
+        isolation. These are already committed to SOMA storage, so they are the unchangeable inputs
+        for a multi-file append-mode ingest to that experiment. The label-to-SOMA-join-ID mappings
+        for the input files will be computed on top of this foundation.
+        """
+
+        obs_map = {}
+        var_maps = {}
+
+        if experiment_uri is None:
+            tiledbsoma.logging.logger.info(
+                "Registration: starting with empty experiment."
+            )
+
+        else:
+            tiledbsoma.logging.logger.info(
+                f"Registration: starting with experiment {experiment_uri}"
+            )
+
+            with tiledbsoma.Experiment.open(experiment_uri) as exp:
+                for batch in exp.obs.read(column_names=["soma_joinid", obs_field_name]):
+                    for i in range(len(batch[0])):
+                        soma_joinid = batch[0][i].as_py()
+                        obs_id = batch[1][i].as_py()
+                        if obs_id not in obs_map:
+                            obs_map[obs_id] = soma_joinid
+
+                for measurement_name in exp.ms:
+                    var_map = {}
+                    for batch in exp.ms[measurement_name].var.read(
+                        column_names=["soma_joinid", var_field_name]
+                    ):
+                        for i in range(len(batch[0])):
+                            soma_joinid = batch[0][i].as_py()
+                            var_id = batch[1][i].as_py()
+                            if var_id not in var_map:
+                                var_map[var_id] = soma_joinid
+                    var_maps[measurement_name] = var_map
+
+            tiledbsoma.logging.logger.info(
+                f"Registration: found nobs={len(obs_map)} nvar={len(var_map)} from experiment."
+            )
+
+        return cls(
+            obs_axis=AxisAmbientLabelMapping(data=obs_map, field_name=obs_field_name),
+            var_axes={
+                ms_name: AxisAmbientLabelMapping(data=data, field_name=var_field_name)
+                for ms_name, data in var_maps.items()
+            },
+        )
+
+    @classmethod
+    def from_anndata_append_on_experiment(
+        cls,
+        adata: ad.AnnData,
+        previous: Self,
+        *,
+        measurement_name: str,
+        obs_field_name: str = "obs_id",
+        var_field_name: str = "var_id",
+    ) -> Self:
+        """Extends registration data to one more AnnData input."""
+        tiledbsoma.logging.logger.info("Registration: registering AnnData object.")
+
+        obs_next_soma_joinid = previous.obs_axis.get_next_start_soma_joinid()
+        obs_map = previous.obs_axis.data
```

Comment:
It looks like you might want to do

```python
obs_map = dict(previous.obs_axis.data)
```

because right now the below is going to end up reaching in and modifying `previous.obs_axis`’s `data` field. (Likewise with other dicts in this factory.)